[DataGridPremium] Automatic parents and children selection (#13757)
mui/mui-x: docs/data/data-grid/row-grouping/row-grouping.md

If you are dynamically switching the `leafField` or `mainGroupingCriteria`, the sorting and filtering models will not be cleaned up automatically, and the sorting/filtering will not be re-applied.
:::

> ## Propagate row selection

Reviewer: ```suggestion
## Automatic children selection
```

Author: I like the second one more.

The motivation behind naming the feature/flag `row selection propagation` is that it not only selects children on selecting parents but also selects, deselects, and puts into indeterminate mode the parents whose children have been selected or deselected. But from an SEO pov, mentioning the simpler use case (children selection) would be better. 👍

By default, selecting a parent row will not select its children.

> Set the `propagateRowSelection` prop to `true` to achieve the following behavior.

Author: I thought about the other names like `groupSelectsChildren`, `smartSelection`, etc. none of which seem to convey everything this prop does. `propagateRowSelection` seemed closer since `stopPropagation` is a widely known phenomenon in the context of DOM events, which is also kinda similar in terms of behavior.

Better suggestions are welcome.

Reviewer 2: For me, this name is fine

If you are dynamically switching the `leafField` or `mainGroupingCriteria`, the sorting and filtering models will not be cleaned up automatically, and the sorting/filtering will not be re-applied.
:::

## Automatic parents and children selection

By default, selecting a parent row does not select its children.
You can override this behavior by using the `rowSelectionPropagation` prop.

Here's how it's structured:

```ts
type GridRowSelectionPropagation = {
  descendants?: boolean; // default: false
  parents?: boolean; // default: false
};
```

When `rowSelectionPropagation.descendants` is set to `true`.

- Selecting a parent would auto-select all its filtered descendants.
- Deselecting a parent row would auto-deselect all its filtered descendants.

When `rowSelectionPropagation.parents` is set to `true`.

- Selecting all the filtered descendants of a parent would auto-select the parent.
- Deselecting a descendant of a selected parent would auto-deselect the parent.

The example below demonstrates the usage of the `rowSelectionPropagation` prop.

> {{"demo": "RowGroupingPropagateSelection.js", "bg": "inline", "defaultCodeOpen": false}}

Reviewer 3: Is this an expected behavior?

https://github.com/user-attachments/assets/457912da-9bcc-42de-9b20-0f95d38cc0d1

Author: Not exactly, `id=0` is failing on `!variable` check.
Fixed: [`c16e7f4` (#13757)](https://github.com/mui/mui-x/pull/13757/commits/c16e7f4b8d60f1a59729555ee8cdad12a0a06eb2)

:::info
The row selection propagation also affects the "Select all" checkbox in a similar way like any other group checkbox.
:::

:::info

> The selected unfiltered rows will be auto-deselected when the filter is applied. Row selection propagation is not applied to the unfiltered rows.

Reviewer 3: ```suggestion
The selected rows are automatically deselected when the filter is applied. Row selection propagation is not applied to the unfiltered rows.
```

> The selected unfiltered rows will be auto-deselected when the filter is applied. Row selection propagation is not applied to the unfiltered rows.

Author: `The selected rows are automatically deselected when the filter is applied` make me understand all the rows would be deselected.
What about adding `that do not pass the filtering criteria` to elaborate more?

```suggestion
The selected rows that do not pass the filtering criteria are automatically deselected when the filter is applied. Row selection propagation is not applied to the unfiltered rows.
```

:::

:::warning

> If `props.disableMultipleRowSelection` is set to `true`, the row selection propagation feature would not work.

Reviewer: ```suggestion
If `props.disableMultipleRowSelection` is set to `true`, the row selection propagation doesn't apply.
```

:::

:::warning

> The row selection propagation is a client-side feature and not recommended to be used with the [server-side data source](/x/react-data-grid/server-side-data/), since it would work on the partially loaded data only.

Reviewer: ```suggestion
Row selection propagation is a client-side feature and is not supported with the [server-side data source](/x/react-data-grid/server-side-data/).
```

:::

> 

Reviewer 2: `indeterminateCheckboxAction` is not mentioned anywhere

Author: We have a section in row selection: https://mui.com/x/react-data-grid/row-selection/#customize-indeterminate-checkbox-behavior

Does replicating it here make sense for better discoverability?

IMO, we might avoid it if we wanna remove that prop in v8 as suggested in https://github.com/mui/mui-x/issues/14300#issuecomment-2313923763

Reviewer 2: No, you can leave it as is then
